File: Assignment1/codes/find_nums.py
"""
    Python code to find the numbers described in the question:
        The product of two consecutive natural numbers which are multiples
        of 3 is equal to 810. Find the two numbers.
"""


from numpy import array, linspace, roots, round, zeros
import matplotlib.pyplot as plt


def my_sum(*args):
    """
    Returns sum of the provided arguments
    """
    s = 0
    for i in args:
        s += i
    return s


def mark_points(points, marker='o', markersize=5, color='black'):
    """
    Marks the given points on the plot and annotates its coordinates

    Parameters
    ----------
    `points`: list of two-tuples to be marked
    `marker`: marker to use
    `markersize`: size of the marker
    `color`: color of the marker
    """

    for i in points:
        plt.plot(i[0], i[1], marker=marker, markersize=markersize, color=color)
        plt.annotate(str(i), i, xytext=(i[0]+0.5, i[1]+20))



def plot(coeffs, roots):
    """
        Plots the given polynomial so that all roots are covered.

        Parameters
        ----------
        `coeffs`: An array of coefficients of polynomial in increasing degree of the variable
        `roots`: An array of roots of polynomial
    """

    x = linspace(1.2*min(roots), 1.2*max(roots), 1000)
    x_axis = zeros(1000)

    temp = [coeffs[i]*x**i for i in range(len((coeffs)))]
    y = array(list(map(my_sum, *temp)))
    plt.grid()
    plt.plot(x, y)
    plt.plot(x, x_axis, color="black")  # x-axis

    y_axis = linspace(1.2*min(y), 1.2*max(y), 1000)
    plt.plot(x_axis, y_axis, color="black")

    points = [(0, 0), *[(i, 0) for i in roots]]
    mark_points(points)

    plt.savefig("../figs/plot.png")
    plt.show()


def find_consecutive_nums(multiple_of, product):
    """
        Parameters
        ----------
        `multiple_of`: The number whose multiple the desired numbers are
        `product`: The product of the two desired numbers

        Returns
        -------
        two consecutive natural numbers which are multiples of `multiple_of`
        and whose product is `product`
    """

    # equation :  (multiple_of)**2*x**2 + (multiple_of)**2*x - (product) = 0

    coefficients = [multiple_of**2, multiple_of**2, -product]

    n1, n2  = roots(coefficients)
    n1, n2 = round((n1, n2))
    n1, n2 = int(n1), int(n2)

    plot(coefficients[::-1], (n1, n2))

    if (n1>0):
        return multiple_of*n1, multiple_of*(n1+1)
    else:
        return multiple_of*n2, multiple_of*(n2+1)


if __name__ == "__main__":
    multiple_of = 3
    product = 810
    n1, n2 = find_consecutive_nums(multiple_of, product)

    try:
        # Verification
        assert n1>0 and n2>0
        assert n1%multiple_of == 0 and n2%multiple_of == 0
        assert n2-n1 == multiple_of
        assert n1*n2 == product
        print(
            f"The consecutive natural numbers n1 and n2 s.t.\n"
            f"n1 % {multiple_of} == 0  and  n2 % {multiple_of} == 0 and n1*n2 == {product} are:\n"
            f"{n1=}\t{n2=}"
        )
    except AssertionError:
        # if verification failed
        print("No such numbers found")

File: Assignment1/codes/find_nums.c
#include <stdio.h>
#include <math.h>
#include <stdbool.h>


#define NUM_POINTS 1000


/*
Stores roots of the equation in root1 and root2.
Returns true if root are real else false.
*/
bool quadratic_roots(double a, double b, double c, double *root1, double *root2){
    double d = b*b - 4*a*c;

    if (d<0)
        return false;

    *root1 = (-b+sqrt(d))/(2*a);
    *root2 = (-b-sqrt(d))/(2*a);

    return true;
}


// write points to be plotted to a file. First line contains roots of equation
void write_points(double a, double b, double c, double root1, double root2){
    FILE *f = fopen("data.txt", "w");

    double min, max;

    if (root1<root2){
        min = root1;
        max = root2;
    }
    else{
        min = root2;
        max = root1;
    }

    
    double x, y, diff;

    min *= 1.2;
    max *= 1.2;

    diff = max-min;

    fprintf(f, "%lf %lf\n", root1, root2);
    

    for (int i=0; i<NUM_POINTS; i++){
        x = min + i*diff/NUM_POINTS;
        y = a*x*x + b*x + c;
        fprintf(f, "%lf %lf\n", x, y);
    }

    fclose(f);
}


/*
Stores the consecutive numbers which are multiples of `multiple_of`
and whose product is `product` in num1 and num2.
Returns true if numbers exist else false.
*/
bool find_cons_num(int product, int multiple_of, int *num1, int *num2){

    double a, b, c, root1, root2;

    // Equation :  multiple_of^2*x^2 + multiple_of^2*x - product = 0
    a = b = pow(multiple_of, 2);
    c = -product;

    bool check = quadratic_roots(a, b, c, &root1, &root2);

    write_points(a, b, c, root1, root2);

    if (!check)
        return false;

    if (root1 > 0)
        *num1 = multiple_of * (int)root1;
    else
        *num1 = multiple_of * (int)root2;

    *num2 = *num1 + multiple_of;

    int n1, n2;
    n1 = *num1;
    n2 = *num2;

    if (n1>0 && n2>0 && n1*n2 == product && n1%multiple_of==0 && n2%multiple_of==0)
        return true;
    return false;
}


int main(){
    int product, multiple_of, num1, num2;

    product = 810;
    multiple_of = 3;

    bool check = find_cons_num(product, multiple_of, &num1, &num2);

    if (!check)
        printf("No such numbers found.\n");
    else
        printf("The numbers are: \n%d %d\n", num1, num2);

    return 0;
}